70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+				shaderSrc->setShaderStage(stage);
+				if (!shaderSrc)
+					return nullptr;
+
+				return m_device->createShader(shaderSrc.get());
+			};
+		auto vs = compileShader("app_resources/fluidParticles.vertex.hlsl", IShader::E_SHADER_STAGE::ESS_VERTEX);
+		auto fs = compileShader("app_resources/fluidParticles.fragment.hlsl", IShader::E_SHADER_STAGE::ESS_FRAGMENT);
+
+		smart_refctd_ptr<video::IGPUDescriptorSetLayout> descriptorSetLayout1, descriptorSetLayout2;
+		{
+			// init descriptors
+			video::IGPUDescriptorSetLayout::SBinding bindingsSet1[] = {
+				{
+					.binding = 0u,
+					.type = asset::IDescriptor::E_TYPE::ET_UNIFORM_BUFFER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+					.stageFlags = asset::IShader::E_SHADER_STAGE::ESS_FRAGMENT,
+					.count = 1u,
+				},
+				{
+					.binding = 1u,
+					.type = asset::IDescriptor::E_TYPE::ET_STORAGE_BUFFER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+					.stageFlags = asset::IShader::E_SHADER_STAGE::ESS_VERTEX,
+					.count = 1u,
+				}
+			};
+			descriptorSetLayout1 = m_device->createDescriptorSetLayout(bindingsSet1);
+			if (!descriptorSetLayout1)
+				return logFail("Failed to Create Render Descriptor Layout 1");
+
+			const auto maxDescriptorSets = ICPUPipelineLayout::DESCRIPTOR_SET_COUNT;
+			const std::array<IGPUDescriptorSetLayout*, maxDescriptorSets> dscLayoutPtrs = {
+				nullptr,
+				descriptorSetLayout1.get(),
+				nullptr,
+				nullptr
+			};
+			m_renderDsPool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::ECF_UPDATE_AFTER_BIND_BIT, std::span(dscLayoutPtrs.begin(), dscLayoutPtrs.end()));
+			m_renderDs = m_renderDsPool->createDescriptorSet(descriptorSetLayout1);
+		}
+
+		// write descriptors
+		{
+			IGPUDescriptorSet::SDescriptorInfo camInfo;
+			camInfo.desc = smart_refctd_ptr(cameraBuffer);
+			camInfo.info.buffer = {.offset = 0, .size = cameraBuffer->getSize()};
+			IGPUDescriptorSet::SDescriptorInfo verticesInfo;
+			verticesInfo.desc = smart_refctd_ptr(particleVertexBuffer);
+			verticesInfo.info.buffer = {.offset = 0, .size = particleVertexBuffer->getSize()};
+			IGPUDescriptorSet::SWriteDescriptorSet writes[2] = {
+				{.dstSet = m_renderDs.get(), .binding = 0, .arrayElement = 0, .count = 1, .info = &camInfo},
+				{.dstSet = m_renderDs.get(), .binding = 1, .arrayElement = 0, .count = 1, .info = &verticesInfo},
+			};
+			m_device->updateDescriptorSets(std::span(writes, 2), {});
+		}
+
+		SBlendParams blendParams = {};
+		blendParams.logicOp = ELO_NO_OP;
+		blendParams.blendParams[0u].srcColorFactor = asset::EBF_SRC_ALPHA;
+		blendParams.blendParams[0u].dstColorFactor = asset::EBF_ONE_MINUS_SRC_ALPHA;
+		blendParams.blendParams[0u].colorBlendOp = asset::EBO_ADD;
+		blendParams.blendParams[0u].srcAlphaFactor = asset::EBF_ONE_MINUS_SRC_ALPHA;
+		blendParams.blendParams[0u].dstAlphaFactor = asset::EBF_ZERO;
+		blendParams.blendParams[0u].alphaBlendOp = asset::EBO_ADD;
```

Comment:
don't set up a blend op if you rely on z-testing, you do `discard` in the shader, the ball is opaque (alpha tested != alpha blended)